log-redaction.md

```diff
@@ -0,0 +1,45 @@
+---
+title: 日志脱敏
+summary: 了解 TiDB 各组件中的日志脱敏
+---
+
+# 日志脱敏
+
+TiDB 在提供详细的日志信息时，可能会把数据库敏感的数据（例如用户数据）打印出来，造成数据安全方面的风险。因此 TiDB、TiKV、PD 等组件各提供了一个配置项开关，开关打开后，会隐藏日志中包含的用户数据值。各组件相关的配置开关如下所示。
+
+例如在以下表中插入用户数据，并在 TiDB 侧开启日志脱敏：
+
+```sql
+mysql> create table t (a int, unique key (a));
+Query OK, 0 rows affected (0.00 sec)
+
+mysql> insert into t values (1),(1);
+ERROR 1062 (23000): Duplicate entry '1' for key 'a'
+mysql> set @@session.tidb_redact_log=1;
+Query OK, 0 rows affected (0.00 sec)
+
+mysql> insert into t values (1),(1);
+ERROR 1062 (23000): Duplicate entry '?' for key 'a'
+```
+
+打印出的错误日志如下：
+
+```
+[2020/10/20 11:45:37.796 +08:00] [INFO] [conn.go:800] ["command dispatched failed"] [conn=5] [connInfo="id:5, addr:127.0.0.1:57222 status:10, collation:utf8_general_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="insert into t values (1),(1)"] [txn_mode=OPTIMISTIC] [err="[kv:1062]Duplicate entry '1' for key 'a'"]
+
+[2020/10/20 11:45:49.539 +08:00] [INFO] [conn.go:800] ["command dispatched failed"] [conn=5] [connInfo="id:5, addr:127.0.0.1:57222 status:10, collation:utf8_general_ci,  user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="insert into t values ( ? ) , ( ? )"] [txn_mode=OPTIMISTIC] [err="[kv:1062]Duplicate entry '?' for key 'a'"]
+```
```

Comment:
```suggestion

可以设置全局的 `tidb_redact_log` 系统变量，来开启 TiDB组件的日志脱敏。
```
mysql> set @@global.tidb_redact_log=1;
```

设置后，所有新的 session 产生的日志都会脱敏：
```sql
mysql> create table t (a int, unique key (a));
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (1),(1);
ERROR 1062 (23000): Duplicate entry '1' for key 'a'
```

打印出的错误日志如下：

```
[2020/10/20 11:45:49.539 +08:00] [INFO] [conn.go:800] ["command dispatched failed"] [conn=5] [connInfo="id:5, addr:127.0.0.1:57222 status:10, collation:utf8_general_ci,  user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="insert into t values ( ? ) , ( ? )"] [txn_mode=OPTIMISTIC] [err="[kv:1062]Duplicate entry '?' for key 'a'"]
```

```